src/measurement.rs

```diff
@@ -0,0 +1,304 @@
+        "failed to read file: {}",
+        vmsa_cpu1_filename.display()
+    ))?;
+
+    let mut ret: Vec<u8> = Vec::new();
+    ret.extend(vmsa_cpu0);
+
+    let mut count = std::cmp::max(num_cpus - 1, 0);
+    while count > 0 {
+        ret.extend(&vmsa_cpu1);
+        count -= 1;
+    }
+
+    Ok(ret)
+}
+
+fn build_digest(args: &BuildArgs, policy: &u32) -> super::Result<Vec<u8>> {
+    if let Some(ld) = &args.launch_digest {
+        return base64::decode(ld).context("failed to base64 decode --launch-digest");
+    }
+
+    if args.firmware.is_none() {
+        return Err(anyhow::anyhow!(
+            "One of --firmware or --launch-digest must be specified."
+        ));
+    }
+
+    let firmware = args.firmware.as_ref().unwrap();
+    let mut content =
+        std::fs::read(firmware).context(format!("failed to read file: {}", &firmware.display()))?;
+
+    content.extend(build_kernel_table(args)?);
+    content.extend(build_cpu_state(args, policy)?);
+
+    sha256_bytes(&content, "firmware + table")
+}
+
+fn parse_hex_or_int(argname: &str, val: &str) -> super::Result<u32> {
+    // Adapted from clap_num crate
+    let result = if val.to_ascii_lowercase().starts_with("0x") {
+        u32::from_str_radix(&val["0x".len()..], 16)
+    } else {
+        val.parse::<u32>()
+    };
+
+    match result {
+        Ok(v) => Ok(v),
+        _ => Err(anyhow::anyhow!(
+            "{}={} value must be int or hex",
+            argname,
+            val
+        )),
+    }
+}
+
+pub fn parse_base64_or_path(argname: &str, val: &str) -> super::Result<Vec<u8>> {
+    let result = if std::fs::metadata(val).is_ok() {
+        std::fs::read(val).context(format!("reading path {}={} failed", argname, val))
+    } else {
+        base64::decode(val).context(format!("failed to base64 decode {}={}", argname, val))
+    };
+
+    match result {
+        Ok(v) => {
+            log::debug!("{} base64: {}", argname, base64::encode(&v));
+            Ok(v)
+        }
+        Err(e) => Err(e),
+    }
+}
+
+pub fn build_cmd(args: BuildArgs) -> super::Result<()> {
+    let mut data: Vec<u8> = Vec::new();
+
+    let policy = parse_hex_or_int("--policy", &args.policy)?;
+
+    let digest = build_digest(&args, &policy)?;
+
+    let api_major = parse_hex_or_int("--api-major", &args.api_major)?;
+    let api_minor = parse_hex_or_int("--api-minor", &args.api_minor)?;
+    let build_id = parse_hex_or_int("--build-id", &args.build_id)?;
+    let tik = parse_base64_or_path("--tik", &args.tik)?;
+
+    let nonce;
+    if let Some(nonce_arg) = args.nonce {
+        nonce = parse_base64_or_path("--nonce", &nonce_arg)?;
+    } else if let Some(launch_measure_blob_arg) = args.launch_measure_blob {
+        let blob = parse_base64_or_path("--launch-measure-blob", &launch_measure_blob_arg)?;
+        nonce = blob[32..].to_vec();
+    } else {
+        return Err(anyhow::anyhow!(
+            "One of --nonce or --launch-measure-blob must be specified."
+        ));
+    }
```

Comment:
Just a suggestion.. if you don't prefer this change then it's no big deal.

```suggestion
    let nonce = if let Some(nonce_arg) = args.nonce {
        parse_base64_or_path("--nonce", &nonce_arg)?;
    } else if let Some(launch_measure_blob_arg) = args.launch_measure_blob {
        let blob = parse_base64_or_path("--launch-measure-blob", &launch_measure_blob_arg)?;
        blob[32..].to_vec();
    } else {
        return Err(anyhow::anyhow!(
            "One of --nonce or --launch-measure-blob must be specified."
        ));
    }
```

Reply:
Hmm. I used github's commit suggestion here but didn't try testing locally, and then CI is failing. Trying to resolve it locally is non-obvious to me, so I dropped the patch and repushed. Maybe handle it as a separate PR when this hits git?

Reply:
Sure, that's fine.